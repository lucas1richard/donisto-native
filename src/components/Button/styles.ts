import { StyleSheet, TextStyle, ViewStyle } from 'react-native';
import theme from 'theme/iftheme';

const defaultViewStyles: ViewStyle = {
  // padding: theme.fontSize,
  height: 46,
  width: '100%',
  borderRadius: 8,
  alignItems: 'center',
  justifyContent: 'center',
  flexDirection: 'row'
};

export const viewStyles = StyleSheet.create({
  outlined: {
    ...defaultViewStyles,
    borderWidth: 1,
    borderColor: theme.primary[600],
    backgroundColor: theme.primary[50]
  },
  rounded: {
    ...defaultViewStyles,
    borderRadius: 99
  },
  contained: {
    ...defaultViewStyles
  },
  text: {
  },
  primary: {
    ...defaultViewStyles,
    backgroundColor: theme.primary[500]
  },
  secondary: {
    ...defaultViewStyles,
    backgroundColor: theme.secondary[400]
  },
  default: {
    ...defaultViewStyles,
    backgroundColor: '#f2f2f2'
  },
  _default: {
    ...defaultViewStyles,
    backgroundColor: '#f2f2f2'
  }
});

const defaultTextStyles: TextStyle = {
  fontSize: theme.fs
};

export const textStyles = StyleSheet.create({
  'text-primary': {
    color: theme.primary[400],
    borderBottomWidth: 1,
    borderBottomColor: theme.primary[400]
  },
  'text-secondary': {
    color: theme.secondary[400]
  },
  'text-default': {
    color: '#000'
  },
  primary: {
    ...defaultTextStyles,
    color: '#fff'
  },
  secondary: {
    ...defaultTextStyles,
    color: '#fff'
  },
  default: {
    ...defaultTextStyles,
    color: '#000'
  },
  _default: {
    ...defaultTextStyles,
    color: '#000'
  }
});
