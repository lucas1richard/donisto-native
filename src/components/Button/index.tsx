import React from 'react';
import { View, ButtonProperties } from 'react-native';
import Touchable from 'components/Touchable';
import { Text } from 'react-native';
import { viewStyles, textStyles } from './styles';
import theme from 'theme/iftheme';

interface IButtonProps extends ButtonProperties {
  title: string;
  onPress: () => any;
  variant?: 'text'|'contained'|'outlined'|'default'|'rounded';
  color?: 'primary'|'secondary'|'default';
  icon?: any
}

class Button extends React.Component<IButtonProps> {
  static defaultProps = {
    color: '_default',
    variant: 'default'
  };

  canPress: boolean = true;
  pressTimer: any;

  get textStyles() {
    const { color, variant } = this.props;
    if (variant === 'text') {
      switch (color) {
        case 'primary':
          return textStyles['text-primary'];
        case 'secondary':
          return textStyles['text-secondary'];
        default:
          return textStyles['text-default'];
      }
    }
    return textStyles[color];
  }

  get viewStyles() {
    const { color, variant } = this.props;
    if (variant === 'text') {
      return viewStyles.text;
    }
    return [viewStyles[color], viewStyles[variant]];
  }

  render() {
    const { onPress, title, icon } = this.props;
    return (
      <Touchable onPress={onPress}>
        <View style={this.viewStyles}>
          {icon ? (
            <View style={{ marginRight: theme.fontSize }}>
              {icon}
            </View>
          ) : undefined}
          <Text style={this.textStyles}>
            {title}
          </Text>
        </View>
      </Touchable>
    );
  }
}

export default Button;
