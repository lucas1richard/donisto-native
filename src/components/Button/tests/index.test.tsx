import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button, { IButtonProps } from '..';

describe('<Button />', () => {
  const renderComponent = (props: IButtonProps) => toJson(shallow(<Button {...props} />));
  it('renders with a title', () => {
    expect(renderComponent({
      title: 'mockTitle',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a primary color', () => {
    expect(renderComponent({
      title: 'mockTitle',
      color: 'primary',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a secondary color', () => {
    expect(renderComponent({
      title: 'mockTitle',
      color: 'secondary',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a text variant', () => {
    expect(renderComponent({
      title: 'mockTitle',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a contained variant', () => {
    expect(renderComponent({
      title: 'mockTitle',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a outlined variant', () => {
    expect(renderComponent({
      title: 'mockTitle',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders with a rounded variant', () => {
    expect(renderComponent({
      title: 'mockTitle',
      variant: 'text',
      onPress: jest.fn()
    })).toMatchSnapshot();
  });
  it('renders a loader', () => {
    expect(renderComponent({
      title: 'mockTitle',
      onPress: jest.fn(),
      showLoader: true,
    })).toMatchSnapshot();
  });
  it('renders with an icon', () => {
    expect(renderComponent({
      title: 'mockTitle',
      onPress: jest.fn(),
      icon: 'mockIcon',
    })).toMatchSnapshot();
  });
});