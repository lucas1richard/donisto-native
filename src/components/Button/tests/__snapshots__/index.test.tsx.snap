// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders a loader 1`] = `<Loader />`;

exports[`<Button /> renders with a contained variant 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "color": "#000",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a outlined variant 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "color": "#000",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a primary color 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "borderBottomColor": "rgb(238, 135, 119)",
          "borderBottomWidth": 1,
          "color": "rgb(238, 135, 119)",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a rounded variant 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "color": "#000",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a secondary color 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "color": "#ffa726",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a text variant 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={Object {}}
  >
    <Component
      style={
        Object {
          "color": "#000",
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with a title 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f2",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 46,
          "justifyContent": "center",
          "width": "100%",
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f2",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 46,
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <Component
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;

exports[`<Button /> renders with an icon 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f2",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 46,
          "justifyContent": "center",
          "width": "100%",
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f2",
          "borderRadius": 8,
          "flexDirection": "row",
          "height": 46,
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <Component
      style={
        Object {
          "marginRight": 20,
        }
      }
    >
      mockIcon
    </Component>
    <Component
      style={
        Object {
          "color": "#000",
          "fontSize": 20,
        }
      }
    >
      mockTitle
    </Component>
  </Component>
</TouchableOpacity>
`;
